### PR TITLE
feat(Host): RHICOMPL-3731 sort systems by num of failed rules

### PR DESCRIPTION
**Pull request by marleystipich2**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/models/host.rb [hunk 1]
```diff
@@ -86,6 +99,13 @@ class Host < ApplicationRecord
     left_outer_joins(:test_result_profiles)
   }
 
+  scope :joins_test_result_profiles_with_failed_rules, lambda {
+    # Needed to GROUP BY Host columns to work correctly with sortable_by :rules_failed
+    left_outer_joins(test_result_profiles: :rule_results).where(
+      rule_results: { result: RuleResult::FAILED }
+    ).group(HOST_COLUMNS)
```
Comment (marleystipich2) [inline app/models/host.rb:106]: Grouping by the host columns was needed for `sortable_by :rules_failed` (like mentioned in the comment) but not sure if this is the best place the add it 🤔

Comment (skateman) [reply app/models/host.rb:106]: A grouping can mess up the object in activerecord in ways that can not be handled by the serializer sometimes. What you could do is to "hide" it in a subquery. However, I would need a bit more context of why the grouping here is needed.

File: app/models/host.rb [hunk 2]
```diff
@@ -53,6 +53,10 @@ class Host < ApplicationRecord
     scope: :with_benchmark
   )
 
+  sortable_by(
+    :rules_failed, 'sq.rules_failed', scope: :with_failed_rules_count
+  )
```
Comment (skateman) [inline app/models/host.rb:58]: This is short enough to be a one-liner without the parentheses

File: app/models/host.rb [hunk 3]
```diff
@@ -86,6 +90,14 @@ class Host < ApplicationRecord
     left_outer_joins(:test_result_profiles)
   }
 
+  scope :with_failed_rules_count, lambda {
+    sq = Host.left_outer_joins(test_result_profiles: :rule_results)
+             .where(rule_results: { result: RuleResult::FAILED })
+             .select(Host.arel_table[:id].as('id'), RuleResult.arel_table[:result].count.as('rules_failed'))
+             .group('hosts.id')
```
Comment (skateman) [inline app/models/host.rb:97]: I think you can omit the `Host` as we are in its context. Maybe not the first one, but the second one for sure, so please try:
```suggestion
    sq = left_outer_joins(test_result_profiles: :rule_results)
             .where(rule_results: { result: RuleResult::FAILED })
             .select(arel_table[:id].as('id'), RuleResult.arel_table[:result].count.as('rules_failed'))
             .group('hosts.id')
```

File: app/models/host.rb [hunk 4]
```diff
@@ -86,6 +90,14 @@ class Host < ApplicationRecord
...
+    joins("INNER JOIN (#{sq.to_sql}) sq ON sq.id = hosts.id")
```
Comment (skateman) [inline app/models/host.rb:98]: So you don't need to select anything? NICE!

**Review by skateman — CHANGES_REQUESTED** [review]
Looks good, just minor cosmetics please

File: app/models/host.rb (hunk 4 above)
Comment (marleystipich2) [reply app/models/host.rb:98]: Right, it works without it 👍

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Pull request merged**